### PR TITLE
FIX: compute_power_by_layer

### LINKED ISSUE #7605: Bug located in compute_power_by_layer
### Before submitting the issue

- [x] I have searched among the existing issues
- [x] I am using a Python virtual environment

### Description of the bug

AEDT is raises an error when computing power loss if the net has no field data on a specific layer. When it happens, PyAEDT will raise an error and release the desktop.

This is correct behavior but not wanted in compute_power_by_layer

### Steps To Reproduce

execute compute_power_by_layer

### Which Operating System are you using?

Windows

### Which Python version are you using?

3.12

### Installed packages

all

### PR DESCRIPTION
**Pull request by hui-zhou-a**
## Description

Add try-except to handle the situation when a net has no field data on the layer. 
- prevent the code from breaking
- set loss to 0

## Issue linked
Close #7605 

## Checklist
- [ ] I have tested my changes locally.
- [ ] I have added necessary documentation or updated existing documentation.
- [ ] I have followed the coding style guidelines of this project.
- [ ] I have added appropriate tests (unit, integration, system).
- [ ] I have reviewed my changes before submitting this pull request.
- [ ] I have linked the issue or issues that are solved by the PR if any.
- [ ] I have agreed with the Contributor License Agreement ([CLA](https://developer.ansys.com/form/cla-acceptance)).

### CODE REVIEW

**Comment by hui-zhou-a** [comment]
close #7605

**Comment by hui-zhou-a** [comment]
To handle this situation, the user need to turn off release_on_exception
```
from ansys.aedt.core import Hfss3dLayout
from ansys.aedt.core.generic.settings import settings

aedt = r"D:\solved_project.aedt"


h3d = Hfss3dLayout(aedt, version="2026.1", non_graphical=True)

settings.release_on_exception = False
loss = h3d.post.compute_power_by_layer()
settings.release_on_exception = True

h3d.release_desktop(False, False)
```

File: src/ansys/aedt/core/visualization/post/post_3dlayout.py [hunk 1]
```diff
@@ -157,9 +158,27 @@ def _compute_power_loss(self, net_filter=None, layer_filter=None, solution: str
...
+                        try:
+                            self._app.post.fields_calculator.add_expression(my_expression, "")
+                            loss = self._app.post.fields_calculator.evaluate(
+                                my_expression["name"], solution, intrinsics={}
+                            )
+                        except Exception:
+                            self.logger.add_warning_message(
+                                f"Net `{net_name}` has no field data on layer `{layer_name}`."
+                            )
+                            loss = 0
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/post/post_3dlayout.py]: I would agree with this pattern (same as `import_from_openstreet_map`) if the second `try:...` snippet was modifying the evaluation method. But that's not the case. Why do you trigger the same code a second time ? Also, what the logic with assigning `loss = 0` ? Multiple issues could arise and lead to this situation.

Please update the PR description.

**Review by SMoraisAnsys — CHANGES_REQUESTED** [review]

File: src/ansys/aedt/core/visualization/post/post_3dlayout.py (hunk 1 above)
Comment (hui-zhou-a) [reply src/ansys/aedt/core/visualization/post/post_3dlayout.py]: The second try is unnecessary and it is removed now. The fix is to workaround AEDT error when there is no field data. In this case, loss is set to 0 instead of breaking. The expression in this method has a fixed pattern, so it is not likely other issue could lead to the exception.

Comment (Samuelopez-ansys) [reply src/ansys/aedt/core/visualization/post/post_3dlayout.py]: @SMoraisAnsys if the layer has several polygons and of them has not power information, AEDT raises an exception, so the fallback is setting these valeus to 0.

Please @hui-zhou-a add a description to the PR.

File: src/ansys/aedt/core/visualization/post/post_3dlayout.py [hunk 2]
```diff
@@ -157,9 +158,21 @@ def _compute_power_loss(self, net_filter=None, layer_filter=None, solution: str
...
 
-                loss = self._app.post.fields_calculator.evaluate(my_expression["name"], solution, intrinsics={})
+                try:
+                    self._app.post.fields_calculator.add_expression(my_expression, "")
+                    loss = self._app.post.fields_calculator.evaluate(my_expression["name"], solution, intrinsics={})
+                except Exception as e:  # pragma: no cover
+                    if settings.release_on_exception:
+                        raise GrpcApiError(
+                            f"Failed to compute power loss for Net `{net_name}` on layer `{layer_name}`."
+                            f"Consider disabling `settings.release_on_exception` for fallback compute. "
```
Comment (SMoraisAnsys) [inline src/ansys/aedt/core/visualization/post/post_3dlayout.py]:
```suggestion
                            f"Failed to compute power loss for Net `{net_name}` on layer `{layer_name}`. "
                            "Set `settings.release_on_exception = False` to fall back to a zero loss value instead of raising."
```

**Review by SMoraisAnsys — DISMISSED** [review]
The changes make more sense now, thanks @hui-zhou-a.
I left a minor comment on the error message.

**Comment by Samuelopez-ansys** [comment]
@hui-zhou-a  Is it possible to add a test?

**Comment by hui-zhou-a** [comment]
> @hui-zhou-a Is it possible to add a test?

Sorry, no. I couldn't make up such a test case. The problem was reported on a customer design.

**Comment by Samuelopez-ansys** [comment]
Hi @hui-zhou-a , this is not a suggestion; we can not add new code that is not tested. You can take the project used for the tests: def test_dcir(dcir_example_project)
And modify it to have a polygons without power.

**Comment by SMoraisAnsys** [comment]
I agree with @Samuelopez-ansys, worst case scenario we can mock the situation and only add a unit test (not system test)

**Comment by hui-zhou-a** [comment]
> Hi @hui-zhou-a , this is not a suggestion; we can not add new code that is not tested. You can take the project used for the tests: def test_dcir(dcir_example_project) And modify it to have a polygons without power.



> Hi @hui-zhou-a , this is not a suggestion; we can not add new code that is not tested. You can take the project used for the tests: def test_dcir(dcir_example_project) And modify it to have a polygons without power.

Tried, but I cannot reproduce the problem on this example board.

**Comment by hui-zhou-a** [comment]
Added test on mock

File: tests/system/layout/test_3dlayout_dcir.py [hunk 3]
```diff
@@ -62,3 +64,15 @@ def test_dcir(dcir_example_project) -> None:
...
+@pytest.mark.skipif(is_linux, reason="Not Supported on Linux.")
+@pytest.mark.skipif(DESKTOP_VERSION == "2025.2", reason="WAITING BUG FIX")
+@patch(
+    "ansys.aedt.core.visualization.post.fields_calculator.FieldsCalculator.add_expression",
+    side_effect=RuntimeError("mock add_expression failure"),
+)
+def test_dcir2(mock_add_expression, dcir_example_project) -> None:
+    assert dcir_example_project.post.compute_power_by_layer()
+    assert mock_add_expression.called
+    assert any(call.args and call.args[-1] == "" for call in mock_add_expression.call_args_list)
```
Comment (SMoraisAnsys) [inline tests/system/layout/test_3dlayout_dcir.py]:
```suggestion
@pytest.mark.skipif(is_linux, reason="Not Supported on Linux.")
@pytest.mark.skipif(DESKTOP_VERSION == "2025.2", reason="WAITING BUG FIX")
@patch(
    "ansys.aedt.core.visualization.post.fields_calculator.FieldsCalculator.add_expression",
    side_effect=RuntimeError("Dummy exception"),
)
def test_dcir_power_loss_fallback_to_zero(mock_add_expression, dcir_example_project) -> None:
    """Test fallback to zero."""
    result = dcir_example_project.post.compute_power_by_layer()

    assert result is not None
    assert all(entry == 0.0 for entry in result.values())
```
The proposed test does not bring much sense in terms of validation. This one should ensure that the fall back is used and the result is the one expected.

**Review by SMoraisAnsys — DISMISSED** [review]

**Comment by hui-zhou-a** [comment]
#7605

**Comment by hui-zhou-a** [comment]
@SMoraisAnsys Could you help me on Labeler check? I referenced this PR in an issue, but it just cannot pass this test.

**Comment by SMoraisAnsys** [comment]
> @SMoraisAnsys Could you help me on Labeler check? I referenced this PR in an issue, but it just cannot pass this test.

Your PR is not up to date so it can't run a job that it doesn't know.

**Comment by Samuelopez-ansys** [comment]
You need to link an issue in the description, if not issue exists, just comment that no issue linked.

**Comment by hui-zhou-a** [comment]
I linked this PR to the issue manually. It should be fine now.

**Review by Samuelopez-ansys — APPROVED** [review]
